### PR TITLE
Improve gRPC connection handling

### LINKED ISSUE #27: Improve reliability of gRPC connections / communications
**Motivation**

I've configured Falco Exporter as a sidecar, which exposes some certain issues, particularly the delay during Falco's BPF probe compilation and gRPC server binding. This causes the exporter to generally go into CrashLoopBackOff due to error.

**Feature**

Implement backoff retry functionality for gRPC client creation, connection and subscription.

```
2020/05/19 00:25:50 exporter error: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp [::1]:5060: connect: connection refused", retrying after backoff of 1 second(s)
2020/05/19 00:25:51 exporter error: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp [::1]:5060: connect: connection refused", retrying after backoff of 2 second(s)
2020/05/19 00:25:53 exporter error: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp [::1]:5060: connect: connection refused", retrying after backoff of 4 second(s)
2020/05/19 00:25:57 exporter error: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp [::1]:5060: connect: connection refused", retrying after backoff of 8 second(s)
```

```
NAME          READY   STATUS    RESTARTS   AGE
falco-6ccmq   2/2     Running   0          58s
falco-hd2sh   2/2     Running   0          49s
falco-s74c5   2/2     Running   0          35s
falco-vm62v   2/2     Running   0          73s
falco-zhtc2   2/2     Running   0          82s
```

**Alternatives**

If I had used the current standard approach of multiple daemonsets/services I could have an initContainer waiting for the service (representing the Falco daemonset) to become available, with a sidecar it is not possible - I'd essentially need an initContainer _only_ for falco-exporter.

### PR DESCRIPTION
**Pull request by leogr**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind bug

> /kind cleanup

> /kind documentation

> /kind tests

/kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

/area commands

/area pkg

> /area examples

**What this PR does / why we need it**:

This PR:
- [x] Enable the built-in gRPC exponential backoff on initial connection only
- [x] Introduce the `--timeout` option (default to 2 mins) for the gRPC initial connection
- [x] Introduce a `/readiness` endpoint that returns 200 only when the gRPC connection is established and everything else is ready
- [x] Update the helm chart and k8s resource accordingly

**Which issue(s) this PR fixes**:

Fixes #27 

**Special notes for your reviewer**:

- retries are performed on initial connection only, it still gives up if the connection is lost later
- `/metrics` is now exposed early to serve as liveness and readiness is handled properly to avoid podcrash loop, that being said if the connection is lost later (while receiving events) the process exits and the container restarts (that is correct IMHO, it would indicate that the Falco gRPC server is down or a serious connection problem)
- blocked until https://github.com/falcosecurity/client-go/pull/44
- #28 will be closed in favor of this PR.
/hold

/cc @fntlnz 
/cc @leodido 

@fujin Could you test this approach and let me know if it solved your issue?
Thanks in advance

**Does this PR introduce a user-facing change?**:

```release-note
update: enable exponential backoff on initial gRPC connection
feat: add `--timeout` option
feat: add readiness endpoint
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,7 +3,11 @@ module github.com/falcosecurity/falco-exporter
 go 1.13
 
 require (
-	github.com/falcosecurity/client-go v0.0.0-20191128160715-8bd88fc0972e
+	// todo(leogr): currently pointing to https://github.com/falcosecurity/client-go/pull/44
+	//              update before merging
+	github.com/falcosecurity/client-go v0.1.1-0.20200525180237-6e2a620c5cfb
```
Comment (leogr) [inline go.mod:8]: Waiting for https://github.com/falcosecurity/client-go/pull/44

**Comment by fujin** [comment]
@leogr tested OK :heavy_check_mark: in my environments. I see the readiness probe failure as anticipated:
```
  Warning  Unhealthy  57s (x8 over 3m57s)  kubelet, gke-xxxxx-xh9v  Readiness probe failed: HTTP probe failed with statuscode: 404
```
after a couple of automatic restarts:
```
2020/05/27 01:36:19 connecting to gRPC server localhost:5060
2020/05/27 01:36:19 listening on :9376/metrics
2020/05/27 01:36:57 connected to gRPC server, subscribing events stream
2020/05/27 01:36:57 ready, receiving events
```

**Comment by leogr** [comment]
Hi @fujin 

Thanks for having tested this! Really appreciated. :+1: 

The `falco-exporter`'s container should restart every 2 mins (ie. the [timeout](https://github.com/falcosecurity/falco-exporter/pull/29/files#diff-47530ed9c9f03a37dcd2ab42f7814c64R28)). So, if you did not want to see any restarts, you can just increase the timeout :)

Out of curiosity, how long does the `falco`'s container take to be up and running? `3m57s`?

Finally, as soon as the [client-go](https://github.com/falcosecurity/client-go/pull/44) and this one get merged, I will release the `v0.2.0`. I believe this fix is a high priority.

Thanks again!

**Comment by leogr** [comment]
Quick update:

We're changing the way [gRPC is communicating](https://github.com/falcosecurity/client-go/tree/feat/bidi-grpc-outputs), that was needed to solve some issues we encountered the current gRPC implementation. This improvement should be released on the next Falco version.

Thus I believe this PR will need to be reworked a bit after the gRPC implementation is getting updated. So I keep this still on hold.

**Comment by leogr** [comment]
Since https://github.com/falcosecurity/client-go/pull/44 was merged and I'm going to add [the Unix socket support](https://github.com/falcosecurity/falco-exporter/issues/30) in another PR, I think we can merge this.

/hold cancel

**Pull request merged**